cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClientCredentials.java

```diff
@@ -92,45 +93,70 @@ public MarketplaceOrderingManager getMarketplaceOrderingManager() {
+        if (AppAuthenticationType.CERTIFICATE.name().equals(credentialView.getAuthenticationType())) {
+            applicationTokenCredentials = certificateBasedCredentials(azureEnvironment);
+        } else {
+            applicationTokenCredentials = secretBasedCredentials(azureEnvironment);
+        }
         AzureTokenCredentials result = applicationTokenCredentials;
-        if (codeGrantFlow.orElse(Boolean.FALSE)) {
-            String refreshToken = credentialView.getRefreshToken();
-            if (StringUtils.isNotEmpty(refreshToken)) {
-                LOGGER.info("Creating Azure credentials for a new delegated token with refresh token, credential: {}", credentialView.getName());
-                String resource = azureEnvironment.managementEndpoint();
-                CBRefreshTokenClient refreshTokenClient = cbRefreshTokenClientProvider.getCBRefreshTokenClient(azureEnvironment.activeDirectoryEndpoint());
-                AuthenticationResult authenticationResult = refreshTokenClient.refreshToken(tenantId, clientId, secretKey, resource, refreshToken, false);
-
-                if (authenticationResult == null) {
-                    String msg = String.format("New token couldn't be obtain with refresh token for credential: %s", credentialView.getName());
-                    LOGGER.warn(msg);
-                    throw new CloudConnectorException(msg);
-                }
 
-                Map<String, AuthenticationResult> tokens = Map.of(resource, authenticationResult);
-                result = new CbDelegatedTokenCredentials(applicationTokenCredentials, resource, tokens, secretKey, authenticationContextProvider,
-                        cbRefreshTokenClientProvider);
-            } else {
-                LOGGER.info("Creating Azure credentials for a new delegated token with authorization code, credential: {}", credentialView.getName());
-                String appReplyUrl = credentialView.getAppReplyUrl();
-                String authorizationCode = credentialView.getAuthorizationCode();
-                result = new CbDelegatedTokenCredentials(applicationTokenCredentials, appReplyUrl, authorizationCode, secretKey, authenticationContextProvider,
-                        cbRefreshTokenClientProvider);
-            }
+        if (credentialView.codeGrantFlow()) {
+            result = delegatedTokenCredentials(azureEnvironment, applicationTokenCredentials);
         } else {
             LOGGER.info("Creating Azure credentials with application token credentials, credential: {}", credentialView.getName());
         }
         return result.withDefaultSubscriptionId(subscriptionId);
     }
 
+    private ApplicationTokenCredentials certificateBasedCredentials(AzureEnvironment azureEnvironment) {
+        LOGGER.info("Creating Azure credentials with certificate and private key: {}", credentialView.getName());
+        String auth = credentialView.getPrivateKeyForCertificate() + credentialView.getCertificate();
```

Comment:
credentialView.getPrivateKeyForCertificate() + credentialView.getCertificate() -> credentialView.getAuth() ?

Reply:
Yes, Azure API is very strange, it expects that the caller concatenates these two things, and Azure lib uses matcher and regexp to separate this.